CancellationException occurs while subscribing to connectable hot source

I think the following scenario introduces a bug.

### Expected behavior

> test
> test

### Actual behavior

> test
> Exception in thread "main" reactor.core.Exceptions$ErrorCallbackNotImplemented: java.util.concurrent.CancellationException: Disconnected
> Caused by: java.util.concurrent.CancellationException: Disconnected
>     at reactor.core.publisher.FluxReplay$ReplaySubscriber.dispose(FluxReplay.java:1202)
>     at reactor.core.publisher.OperatorDisposables.dispose(OperatorDisposables.java:132)
>     at reactor.core.publisher.FluxRefCount$RefCountMonitor.innerCancelled(FluxRefCount.java:132)
>     at reactor.core.publisher.FluxRefCount$RefCountInner.cancel(FluxRefCount.java:200)
>     at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:75)
>     at reactor.core.publisher.FluxRefCount$RefCountInner.onNext(FluxRefCount.java:177)
>     at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replayNormal(FluxReplay.java:808)
>     at reactor.core.publisher.FluxReplay$SizeBoundReplayBuffer.replay(FluxReplay.java:892)
>     at reactor.core.publisher.FluxReplay.subscribe(FluxReplay.java:1085)
>     at reactor.core.publisher.FluxRefCount$RefCountMonitor.subscribe(FluxRefCount.java:116)
>     at reactor.core.publisher.FluxRefCount.subscribe(FluxRefCount.java:77)
>     at reactor.core.publisher.MonoNext.subscribe(MonoNext.java:40)
>     at reactor.core.publisher.Mono.subscribe(Mono.java:3077)
>     at reactor.core.publisher.Mono.subscribeWith(Mono.java:3185)
>     at reactor.core.publisher.Mono.subscribe(Mono.java:3071)
>     at reactor.core.publisher.Mono.subscribe(Mono.java:3038)
>     at reactor.core.publisher.Mono.subscribe(Mono.java:2985)
>     at test.Test.main(Test.java:10)

### Steps to reproduce
```
Flux<String> flux = Flux.<String>create(sink -> sink.next("test"))
    .replay(1)
    .refCount();

flux.subscribe(System.out::println);
flux.next().subscribe(System.out::println); // The exception is thrown here!
```
### Reactor Core version
[3.1.4](https://mvnrepository.com/artifact/io.projectreactor/reactor-core/3.1.4.RELEASE)
### JVM version (e.g. `java -version`)
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

I also posted a question to [Stackoverflow](https://stackoverflow.com/questions/48888858/cancellationexception-occurs-while-subscribing-to-connectable-hot-source).

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxRefCount.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxRefCount.java
@@ -113,9 +113,11 @@ final class FluxRefCount<T> extends Flux<T> implements Scannable, Fuseable {
 			// FIXME think about what happens when subscribers come and go below the connection threshold concurrently
 
 			RefCountInner<T> inner = new RefCountInner<>(s, this);
+
+			int subCount = SUBSCRIBERS.incrementAndGet(this);
 			parent.source.subscribe(inner);
-			
-			if (SUBSCRIBERS.incrementAndGet(this) == n) {
+
+			if (subCount == n) {
 				parent.source.connect(this);
 			}
 		}

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxRefCountTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxRefCountTest.java
@@ -45,6 +45,41 @@ public class FluxRefCountTest {
 	}*/
 
 	@Test
+	public void cancelDoesntTriggerDisconnectErrorOnFirstSubscribeNoComplete() {
+		AtomicInteger nextCount = new AtomicInteger();
+		AtomicReference<Throwable> errorRef = new AtomicReference<>();
+		Flux<String> flux = Flux.<String>create(sink -> {
+			sink.next("test");
+		})
+				.replay(1)
+				.refCount(1);
+
+		flux.subscribe(v -> nextCount.incrementAndGet(), errorRef::set);
+		flux.next().subscribe(v -> nextCount.incrementAndGet(), errorRef::set);
+
+		assertThat(nextCount).hasValue(2);
+		assertThat(errorRef).hasValue(null);
+	}
+
+	@Test
+	public void cancelDoesntTriggerDisconnectErrorOnFirstSubscribeDoComplete() {
+		AtomicInteger nextCount = new AtomicInteger();
+		AtomicReference<Throwable> errorRef = new AtomicReference<>();
+		Flux<String> flux = Flux.<String>create(sink -> {
+			sink.next("test");
+			sink.complete();
+		})
+				.replay(1)
+				.refCount(1);
+
+		flux.subscribe(v -> nextCount.incrementAndGet(), errorRef::set);
+		flux.next().subscribe(v -> nextCount.incrementAndGet(), errorRef::set);
+
+		assertThat(nextCount).hasValue(2);
+		assertThat(errorRef).hasValue(null);
+	}
+
+	@Test
 	public void normal() {
 		EmitterProcessor<Integer> e = EmitterProcessor.create();
 

